[FLINK-21190][runtime-web] Expose exception history

XComp opened the pull request:
## What is the purpose of the change

The purpose of this PR is to expose the exception history work that was introduced in the preceding tasks of [FLINK-21187](https://issues.apache.org/jira/browse/FLINK-21187) and [FLINK-21188](https://issues.apache.org/jira/browse/FLINK-21188). 

[FLINK-21189](https://issues.apache.org/jira/browse/FLINK-21189) as an additional preceding task to prepare the work on [FLINK-21190](https://issues.apache.org/jira/browse/FLINK-21190) was postponed due to other higher-prioritized work. Lacking the work on FLINK-21189 and @vthinkxie being busy on other tasks till April, we decided to change the scope of FLINK-21190 to only cover the root causes in the exception history. I created [FLINK-21867](https://issues.apache.org/jira/browse/FLINK-21867) to cover the follow-up work on the UI.

## Brief change log

* Introduced new parameter `jobmanager.exception-history-size` to handle the size of the internally collected most recent failures
* `BoundedFIFOQueue` was introduced to have a limited amount of failures collected in memory. The `BoundedFIFOQueue` is initialized in the `SchedulerBase` constructor using the newly introduced config option `jobmanager.exception-history-size`.
* Extended `JobExceptionsInfo` by `JobExceptionsInfoWithHistory` which covers the new REST response interface in contrast to the (now deprecated) members of `JobExceptionsInfo`
* `JobExceptionsHandler` is now utilizing `JobExceptionsInfoWithHistory` introducing the new field `exceptionHistory` in the REST response.
* `ArchivedTaskManagerLocation` was introduced to have a static version of `TaskManagerLocation` to overcome caching issues in the `InetAddress` instance of `TaskManagerLocation`. A class was introduced in contrast to just picking the necessary members as the selection of data that should be exposed through the REST API should be done in the `JobExceptionsHandler` (not in the `SchedulerBase` code).

other hotfixes:
* The `ExecutionGraphInfo` now uses the `AccessExecutionGraph` interface instead of returning `ArchivedExecutionGraph`
* Refactored `JsonArchivist` introducing a separate interface `OnlyExecutionGraphJsonArchivist` for REST handlers that don't need to bother about `ExecutionGraphInfo`

## Verifying this change

* `JobExceptionsHandlerTest` got extended testing the new functionality
* `DefaultSchedulerTest` was extended to cover the truncation of exception history entries
* `BoundedFIFOQueueTest` was added to test `BoundedFIFOQueue`'s functionality

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs (REST API)

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/history/OnlyExecutionGraphJsonArchivist.java:
```diff
@@ -0,0 +1,53 @@
+ * Interface for all classes that want to participate in the archiving of job-related json responses
+ * but only provide {@link AccessExecutionGraph}-related information.
+ */
+public interface OnlyExecutionGraphJsonArchivist extends JsonArchivist {
```
This fixes the `ExecutionGraphInfo` bad interface that was introduced in [FLINK-21188](https://issues.apache.org/jira/browse/FLINK-21188): `JsonArchivist` should not implement two methods as it will always rely. on `ExecutionGraphInfo`. This new interface `OnlyExecutionGraphJsonArchivist` is introduced to cover `AbstractAccessExecutionGraphHandler` implementation which don't care about `ExecutionGraphInfo`

zentol replied:
Why are we bothering with this distinction in the first place? Any archivist working on the `ExecutionGraphInfo` has access to the `AccessExecutionGraph` anyway. It should be trivial to adjust existing `JsonArchivist` implementations accordingly.

XComp replied:
The idea is to not bother about the `ExecutionGraphInfo`  in the implementations like the implementing classes do by deriving from `AbstractAccessExecutionGraphHandler`

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java:
```diff
@@ -79,26 +113,35 @@ public void testGetJobExceptionsInfo() throws HandlerRequestException {
-        return new ArchivedExecutionGraphBuilder().setTasks(tasks).build();
+
+        final SortedSet<TaskFailureHistoryEntry> exceptionHistory = new TreeSet<>();
+        exceptionHistory.add(
```
This was added temporarily to have the test succeed again. It's going to be cleaned up as part of the test refactoring.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobExceptionsInfo.java:
```diff
@@ -100,6 +121,11 @@ public boolean isTruncated() {
     }
 
+    @JsonIgnore
+    public List<JobExceptionHistoryEntry> getExceptionHistory() {
```
Looking at it, I guess, we would have to introduce another class that holds the list and a `truncated` flag.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskFailureHistoryEntry.java:
```diff
@@ -0,0 +1,77 @@
+
+    public static TaskFailureHistoryEntry createGlobalFailureEntry(
+            Throwable failureCause, long timestamp) {
+        return new TaskFailureHistoryEntry(new ErrorInfo(failureCause, timestamp), null, null);
```
Seems like `TaskFailureHistoryEntry` is not an appropriate name? Why not just extend the `ErrorInfo` instead?

XComp replied:
I renamed it `ExceptionHistoryEntry`. I hesitated to merge `ErrorInfo` and `ExceptionHistoryEntry` because I wanted to have the exception history decoupled from the `Execution`? The `Execution` only needs to know about the failure and the timestamp.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java:
```diff
@@ -87,18 +87,23 @@ protected JobExceptionsInfo handleRequest(
-            AccessExecutionGraph executionGraph, int exceptionToReportMaxSize) {
+            ExecutionGraphInfo executionGraphInfo, int exceptionToReportMaxSize) {
+        final ArchivedExecutionGraph executionGraph =
+                executionGraphInfo.getArchivedExecutionGraph();
```
Why is the `ExecutionGraphInfo` storing/returning an `ArchivedExecutionGraph` instead of an `AccessExecutionGraph`?

XComp replied:
Fair point. I missed that when adding the `ExecutionGraphInfo` because it was replacing the `ArchivedExecutionGraph`. I changed it.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobExceptionsInfo.java:
```diff
@@ -156,4 +182,56 @@ public int hashCode() {
+
+    /** Nested class to hold exception information of previous failures. */
+    public static final class JobExceptionHistoryEntry {
+        public static final String FIELD_NAME_ROOT_EXCEPTION = "root-exception";
+        public static final String FIELD_NAME_TIMESTAMP = "timestamp";
+        public static final String FIELD_NAME_ALL_EXCEPTIONS = "all-exceptions";
+        public static final String FIELD_NAME_TRUNCATED = "truncated";
```
re-use existing constants instead

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobExceptionsInfo.java:
```diff
@@ -156,4 +182,56 @@ public int hashCode() {
     }
+
+    /** Nested class to hold exception information of previous failures. */
+    public static final class JobExceptionHistoryEntry {
```
we could move this to the upper level, and have JobExceptionsInfo extend this class; then we wouldn't duplicate the entire thing.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -775,9 +775,14 @@ public final void notifyPartitionDataAvailable(final ResultPartitionID partition
     @VisibleForTesting
     protected List<TaskFailureHistoryEntry> getExceptionHistory() {
-        return taskFailureHistory;
+        return new ArrayList<>(taskFailureHistory);
```
Why do you need a copy, and why would wrapping it in an unmodifiable list not suffice?

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java:
```diff
@@ -104,8 +104,12 @@ protected JobExceptionsInfo handleRequest(
             ExecutionGraphInfo executionGraphInfo, int exceptionToReportMaxSize) {
+        final List<TaskFailureHistoryEntry> rootCauseFirstExceptionHistory =
+                executionGraphInfo.getExceptionHistory();
+        Collections.reverse(rootCauseFirstExceptionHistory);
```
I feel like such business should be handled in the UI.

XComp replied:
You have a point. It's just that reversing the order makes it easier to select the root cause which is the last element.

XComp replied:
This is resolved after introducing the `EvictingBoundedList` internally.

XComp commented:
I added another column. The current UI version looks like this
<img width="1526" alt="Screenshot 2021-03-04 at 19 09 18" src="https://user-images.githubusercontent.com/1101012/110009841-dcfca300-7d1d-11eb-98a5-fefe9573c380.png">

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/util/BoundedFIFOQueueTest.java:
```diff
@@ -0,0 +1,100 @@
+    public void testEmptyQueue() {
+        final BoundedFIFOQueue<Integer> testInstance = new BoundedFIFOQueue<>(0);
+        assertThat(testInstance, iterableWithSize(0));
+        testInstance.add(1);
+        assertThat(testInstance, iterableWithSize(0));
```
Fair enough, I changed the test method's name.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/util/BoundedFIFOQueueTest.java:
```diff
@@ -0,0 +1,100 @@
+        final BoundedFIFOQueue<Integer> testInstance = new BoundedFIFOQueue<>(0);
+        assertThat(testInstance.getElementCount(), is(0));
+
+        testInstance.add(1);
+        assertThat(testInstance.getElementCount(), is(1));
```
Does it? Isn't it clear enough through the JavaDoc how the two methods behave? I added further JavaDoc to the test class to make this clearer.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/util/BoundedFIFOQueueTest.java:
```diff
@@ -0,0 +1,100 @@
+        final BoundedFIFOQueue<Integer> testInstance = new BoundedFIFOQueue<>(0);
+        assertThat(testInstance.size(), is(0));
+
+        testInstance.add(1);
+        assertThat(testInstance.size(), is(0));
```
Fair enough, I added another test and renamed the method accordingly

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java:
```diff
@@ -132,7 +143,61 @@ private static JobExceptionsInfo createJobExceptionsInfo(
+
+        return new JobExceptionsInfoWithHistory.JobExceptionHistory(
+                exceptionHistoryEntries,
+                historyEntries.getElementCount(),
```
The exception count gives an understanding on how many exceptions where caught in total since we're limiting the history through a parameter. Thinking about it once more: It might not be useful for long-running jobs, does it?

XComp replied:
My initial motivation was to make it more transparent that the exceptions listed are just the most recent ones. Removing the `total` count from the response object might make the user assume that these are the only exceptions that were caught.

XComp replied:
What about renaming `exception-history` into `most-recent-exceptions`, instead?

zentol replied:
My immediate assumption when hearing "recent" is that it depends on time in some form, which doesn't apply here.
I think "exception-history" is fine.

XComp replied:
I had a chat with @NicoK about it. He agrees that (on the UI side) it's not necessarily useful to have the over all count of exceptions presented as it might grow for long-running jobs. Instead, we could introduce a `max-size-reached` flag to indicate that exception history's maximum size was reached. The UI would then just print `...` and have a refer to the corresponding configuration option and the logs for further analysis.

XComp replied:
The truncation information is not exposed anymore to avoid confusion. A static info text was added that pops up as soon as the history has some entries.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobExceptionsInfo.java:
```diff
@@ -70,14 +80,32 @@ public boolean equals(Object o) {
+
+    @Override
+    public String toString() {
+        return toStringHelper(this)
```
To be honest, that is the generated version of Intellij. I considered it to be better than the `StringBuilder` or String concatenation approaches. But I overlooked the `StringJoiner` approach which is usable in Java8 and uses native Java code. I switched to that version.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/BoundedFIFOQueue.java:
```diff
@@ -0,0 +1,141 @@
+     * @throws NullPointerException If {@code null} is passed as an element.
+     */
+    public void add(T element) {
+        elementCount++;
```
args, good catch. I fixed it

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/ExceptionHistoryEntry.java:
```diff
@@ -0,0 +1,78 @@
+        return new ExceptionHistoryEntry(cause, timestamp, null, null);
+    }
+
+    public static ExceptionHistoryEntry fromFailedExecution(Execution execution) {
```
True 🤦 I applied your suggestion.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/BoundedFIFOQueue.java:
```diff
@@ -0,0 +1,141 @@
+        this(initialElements.size(), initialElements, initialElements.size());
+    }
+
+    private BoundedFIFOQueue(int maxSize, Iterable<T> initialElements, int initialElementCount) {
+        Preconditions.checkArgument(maxSize >= 0, "The maximum size should be at least 0.");
+
+        this.maxSize = maxSize;
+        this.elements = new LinkedList<>();
+        initialElements.forEach(elements::add);
+        this.elementCount = initialElementCount;
+    }
```
We would lose the `elementCount` when copying a `BoundedFIFOQueue` in cases where `elementCount>maxSize`.

XComp replied:
The `elementCount` was reverted. Hence, this doesn't apply anymore.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/BoundedFIFOQueue.java:
```diff
@@ -0,0 +1,141 @@
+            private static final long serialVersionUID = -2763611608109517436L;
+
+            @Override
+            public void add(E element) {
```
You're right. Returning an `Iterable` would be the easiest solution if we decide that the total exception value returned as part of the exception history is not a feature we want. Otherwise, we would have to introduce an interface.

I will wait with the resolution after the `total` discussion is sorted.

XComp replied:
This was removed after we reverted `elementCount`. We're relying on `Iterable` again.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java:
```diff
@@ -37,68 +36,258 @@
+                response,
+                containsGlobalFailure(rootCause.getException(), rootCause.getTimestamp()));
+        assertThat(response, hasHistoryEntryCount(1));
+        assertThat(response, not(hasTruncatedHistory()));
```
I see what you're saying. The `truncated` value is solely used by frontend to determine whether it can get even more entries by increasing the `GET` parameter `maxExceptions`. One way to resolve that would be to rename the `truncate` flag into something more meaningful. This comes with the problem that we would have to do a larger refactoring since we don't want to break the existing REST response schema.

But I'm gonna leave the resolution of this issue for now as it would not be an issue if [we decide that the total exception count is not a feature we want](https://github.com/apache/flink/pull/15049#discussion_r590203591).

XComp replied:
This test was removed as part of the `elementCount` revert

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/BoundedFIFOQueue.java:
```diff
@@ -102,11 +102,12 @@ private BoundedFIFOQueue(int maxSize, Iterable<T> initialElements, int initialEl
     public void add(T element) {
-        elementCount++;
-        elements.add(Preconditions.checkNotNull(element));
+        if (elements.add(Preconditions.checkNotNull(element))) {
```
we don't really need the condition. Even if we were to use some other data-structure in the future where the insertion might fail, based on the current semantics of elementCount it should be incremented in any case.

XComp replied:
This was refactored as part of the `elementCount` revert

XComp commented:
<img width="1282" alt="Screenshot 2021-03-11 at 15 49 52" src="https://user-images.githubusercontent.com/1101012/110815457-be9d2700-828a-11eb-9566-cf06c15f93b8.png">
<img width="1282" alt="Screenshot 2021-03-11 at 15 49 22" src="https://user-images.githubusercontent.com/1101012/110815444-ba710980-828a-11eb-8f3e-3aee4953c841.png">
<img width="1282" alt="Screenshot 2021-03-11 at 15 49 33" src="https://user-images.githubusercontent.com/1101012/110815453-be049080-828a-11eb-875c-d7315dc4b9a0.png">

The `elementCount` was reverted. The UI info message was replaced by a static one that is shown as soon as there are exceptions listed.

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java:
```diff
@@ -37,68 +36,231 @@
+
+        final Throwable failureCause = new RuntimeException("root cause");
+        final long failureTimestamp = System.currentTimeMillis();
+        final List<ExceptionHistoryEntry> exceptionHistory = new ArrayList<>();
```
could use Arrays.asList

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java:
```diff
@@ -154,4 +342,285 @@ private static ArchivedExecutionJobVertex createArchivedExecutionJobVertex(
+            this.genericExpectedMessage = stringBuilder.toString().trim();
+        }
+
+        @Override
```
why are we overriding this method here?

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobExceptionsInfoWithHistory.java:
```diff
@@ -0,0 +1,162 @@
+    public static final class JobExceptionHistory {
+
+        public static final String FIELD_NAME_ENTRIES = "entries";
+        public static final String FIELD_NAME_MAX_SIZE_REACHED = "max-size-reached";
```
Damn, that one I missed because initially, I thought that we could access the configuration from within the `JobExceptionsHandler`

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/ExceptionHistoryEntryTest.java:
```diff
@@ -0,0 +1,176 @@
+     * @return The {@code TestingExecution} instance.
+     */
+    private static TestingExecution createExecutionWithoutFailure() {
+        return new TestingExecution(null, new LocalTaskManagerLocation());
```
Fair enough. Firstly, I wanted to oppose this because I think that the method name helps understanding what the constructor call does. But using a proper local variable name is good enough, I guess.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/ExceptionHistoryEntryTest.java:
```diff
@@ -0,0 +1,176 @@
+            Throwable failureCause,
+            long failureTimestamp,
+            TaskManagerLocation taskManagerLocation) {
+        return new TestingExecution(
```
same as above

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -667,16 +672,15 @@ protected final void archiveFromFailureHandlingResult(
-                            });
+            final ExceptionHistoryEntry exceptionHistoryEntry =
+                    ExceptionHistoryEntry.fromFailedExecution(
+                            executionOptional.get(),
+                            executionOptional.get().getVertex().getTaskNameWithSubtaskIndex());
```
🤦

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobExceptionsInfoWithHistory.java:
```diff
@@ -0,0 +1,162 @@
+        }
+
+        @Override
+        public int hashCode() {
```
You're right. The `RestResponseMarshallingTestBase` asserts the unmarshalled object using [equals](https://github.com/apache/flink/blob/c6997c97c575d334679915c328792b8a3067cfb5/flink-runtime/src/test/java/org/apache/flink/runtime/rest/messages/RestResponseMarshallingTestBase.java#L37). I included `hashCode` for the sole purpose of complying to the `equals/hashCode` contract. I still prefer keeping it like that for consistency reasons, though.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/ExceptionHistoryEntry.java:
```diff
@@ -0,0 +1,95 @@
+    private static final long serialVersionUID = -3855285510064263701L;
+
+    @Nullable private final String failingTaskName;
+    @Nullable private final TaskManagerLocation assignedResourceLocation;
```
Interesting finding. I'm wondering whether it would make sense to create a `TaskManagerLocation` interface since we have the same issue with the `ArchivedExecutionVertex`.

XComp replied:
I created a `ArchivedTaskManagerLocation` for now as the refactoring would take a larger effort which is out of scope for this ticket. @zentol Do we already have a ticket that addresses this refactoring? If not, whats the headline of the refactoring ticket I would create? I understood that introducing `TaskManagerLocation` interface was part of a larger effort?

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java:
```diff
@@ -132,7 +142,62 @@ private static JobExceptionsInfo createJobExceptionsInfo(
+                historyEntry.getException().getOriginalErrorClassName(),
+                historyEntry.getExceptionAsString(),
+                historyEntry.getTimestamp(),
+                Collections.singletonList(
```
You have a good point here. I refactored the schema not relying on the old structure anymore. The refactoring also covered your concern. The `taskName` and `location` are not shared anymore if `null`

XComp commented:
I addressed your comments, squashed and rebased the branch

zentol commented on flink-runtime-web/web-dashboard/src/app/pages/job/exceptions/job-exceptions.component.html:
```diff
@@ -31,6 +31,7 @@
         <tr>
           <th nzShowExpand></th>
           <th>Time</th>
+            <th>Exception</th>
```
indentation

zentol commented on flink-runtime-web/web-dashboard/src/app/pages/job/exceptions/job-exceptions.component.html:
```diff
@@ -40,21 +41,28 @@
           <tr>
             <td nzShowExpand [(nzExpand)]="exception.expand"></td>
             <td>{{exception.timestamp | date:'yyyy-MM-dd HH:mm:ss'}}</td>
+              <td><div class="name">{{exception.exceptionName}}</div></td>
```
indentation

zentol commented on flink-runtime-web/web-dashboard/src/app/pages/job/exceptions/job-exceptions.component.html:
```diff
@@ -40,21 +41,28 @@
           </tr>
         </ng-container>
+        <tr *ngIf="listOfException.length > 0">
+            <td colspan="6">
+                <i nz-icon nzType="info-circle" nzTheme="fill"></i>&nbsp;
+                <i>The exception history is limited to the most recent failures that caused parts of the job or the entire job to restart. The maximum history size can be configured through the Flink configuration.</i>
+            </td>
```
indentation is different to remaining file

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobExceptionsInfoWithHistory.java:
```diff
@@ -0,0 +1,273 @@
+        public ExceptionInfo(
+                @JsonProperty(FIELD_NAME_EXCEPTION_NAME) String exceptionName,
+                @JsonProperty(FIELD_NAME_EXCEPTION_STACKTRACE) String stacktrace,
+                @JsonProperty(FIELD_NAME_EXCEPTION_TIMESTAMP) Long timestamp,
```
That's a copy&paste error coming from `JobExceptionsInfo` I have to admit. You're right - we can use the primitive type here.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandler.java:
```diff
@@ -132,7 +142,75 @@ private static JobExceptionsInfo createJobExceptionsInfo(
+    static String toString(@Nullable ExceptionHistoryEntry.ArchivedTaskManagerLocation location) {
+        return location != null
+                ? taskManagerLocationToString(location.getFQDNHostname(), location.getPort())
+                : null;
```
Good idea 👍

bytesandwich commented:
Hey @XComp would you mind chatting through the change - maybe in a quick call?  I'm new to Flink and I've go the followup FLINK-21439 ticket (it will be my second ticket in Flink, though I'm not a novice developer generally). It might make sense to do before merging this in case you decide to make any changes.

XComp commented:
Hi @bytesandwich , it would be great if you could share your thoughts on the code in this PR instead of discussing it in a call. This makes the decision making process more transparent.

bytesandwich reviewed (changes requested):
Hi @XComp No problem. This looks like a nice change :) Could you please help make the context clearer to me as someone coming fresh to this PR?

It seems that the purpose of this comment is outlined most clearly in this comment:
https://issues.apache.org/jira/browse/FLINK-21190?focusedCommentId=17284745&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17284745

If that is the general purpose of this change, could you please update:

- [ ]  the description 
- [ ] change log
- [ ] verification
 
 ... at the root of this PR as you did so well with https://github.com/apache/flink/pull/14798 ? I think you would probably do this before completing the PR anyway, and this would help me understand your intent with these changes.

zentol commented:
> Could you please help make the context clearer to me as someone coming fresh to this PR?

The goal of this PR and FLINK-21190 is to allow the DefaultScheduler (opposed to the AdaptiveScheduler that is relevant to you in FLINK-21439) to remember prior failure reasons. Currently the DefaultScheduler only remembers the very last failure cause, which for long-term deployments is of course undesirable.
The SchedulerBase now contains a BoundedFIFOQueue containing up to N exceptions that previously caused job restarts; this data-structure is populated by the scheduler.

Contrary to what you might expect, this history is not stored in the ExecutionGraph, which is more-or-less the core data-structure of the job. This is in preparation for FLINK-21439. 
For the DefaultScheduler there only ever exists 1 ExecutionGraph; it is created when the job starts and destroyed when it completes or the JobMaster looses leadership. This is very different for the AdaptiveScheduler, which re-creates the ExecutionGraph repeatedly during execution of a job, specifically whenever it re-scales the job to run at a different parallelism.
Hence, this information must not be stored in the ExecutionGraph. That is also why the `ExecutionGraphInfo` exists; to provide some access point for information about jobs that is not tied to the lifecycle of the ExecutionGraph.

Other changes include various changes to the REST handlers to extract exceptions from the EGInfo object, as well as changes to the UI.

Testing wise this issue can be tested manually quite easy; start a cluster and submit a job that repeatedly fails (imaging a map function throwing an exception every N elements). Additional unit tests were added for all newly added components, and the DefaultSchedulerTest was extended to cover the history mechanism.


I hope this helps you out a bit.


I will leave this PR open for a few more days for further discussions, but please keep in mind that currently I'd be fine with merging it as is and don't want to delay that too much. We can always some things later in FLINK-21439 if there is a need for it.

XComp commented:
Thanks @bytesandwich for pointing out that I haven't updated the PR description, yet. I should have done it earlier. And thanks @zentol for jumping in and explaining the context of this change.

I updated the PR's description and created a follow-up Jira issue [FLINK-21867](https://issues.apache.org/jira/browse/FLINK-21867) to cover the parts that were actually planned for FLINK-21190 but then postponed. I also added explainations for that in the corresponding tickets.

Let us know if you have any further questions. It might make sense to move the discussion into [FLINK-21439](https://issues.apache.org/jira/browse/FLINK-21439], though, as @zentol suggested. ...if you have no questions about this PR specifically.

bytesandwich commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/ExecutionGraphInfo.java:
```diff
@@ -35,15 +33,15 @@
 
-    private final ArchivedExecutionGraph executionGraph;
-    private final List<ErrorInfo> exceptionHistory;
+    private final AccessExecutionGraph executionGraph;
```
Is `ExecutionGraphInfo` still `Serializable` if it can contain an `AccessExecutionGraph` which is an unserializable `DefaultExecutionGraph`? Is [FileExecutionGraphInfoStore](https://github.com/XComp/flink/blob/e284963a18df3004b6b70fbcf38f1682c11d8584/flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/FileExecutionGraphInfoStore.java#L279) going to throw a  ` NotSerializableException` when it tries to serialize such an `ExecutionGraphInfo`?

XComp replied:
That's actually a good point I haven't considered when refactoring this piece of code. Theoretically it should not happen that a `DefaultExecutionGraph` is being returned by `ExecutionGraphInfo`. But it should reflected by the used interface as well. I'm gonna go ahead and introduce a `SerializableAccessExecutionGraph` @zentol do you see any problems with that approach?

zentol replied:
I suppose that would be an argument to stick with the ArchivedExecutionGraph instead.

zentol replied:
I don't see a lot of value in adding another interface for the sole purposes of differentiating between the 2 current implementations.

XComp replied:
I thought of keeping the interface considering that we might have to come up with something new for `AdaptiveScheduler`. But I'm also happy to revert the changes since we could argue that this would be out of scope for this ticket.

zentol replied:
Let's revert it for now and re-evaluate things down the line.

bytesandwich commented:
Thank you very much for the description @zentol! Also thank you @XComp for updating the description. I really appreciate your effort to help me understand :) I will try to pay it forward! I had one question about serializing an `ExecutionGraphInfo` that contains a `DefaultExecutionGraph` and I included it inline. PTAL @XComp . 

My other question, @XComp  @zentol, if AdaptiveScheduler re-creates the ExecutionGraph repeatedly during execution of a job will we have one `JobID` with multiple `ExecutionGraphInfo`s then? And then each of the `ExecutionGraphInfo`s will have multple exceptions? Will we need to make [ExecutionGraphStore.get(jobId)](https://github.com/XComp/flink/blob/b7e93fd671a0dc0e7ba919a447ca9e51bb451fdd/flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/ExecutionGraphInfoStore.java#L49) return a list of `ExecutionGraphInfo`s (and all the other similar changes)?

It seems like this PR's REST API otherwise will work with one job -> many ExecutionGraphInfo -> many exceptions? Would that work in the GUI?

XComp commented:
> My other question, @XComp @zentol, if AdaptiveScheduler re-creates the ExecutionGraph repeatedly during execution of a job will we have one `JobID` with multiple `ExecutionGraphInfo`s then? And then each of the `ExecutionGraphInfo`s will have multple exceptions? Will we need to make [ExecutionGraphStore.get(jobId)](https://github.com/XComp/flink/blob/b7e93fd671a0dc0e7ba919a447ca9e51bb451fdd/flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/ExecutionGraphInfoStore.java#L49) return a list of `ExecutionGraphInfo`s (and all the other similar changes)?
> 
> It seems like this PR's REST API otherwise will work with one job -> many ExecutionGraphInfo -> many exceptions? Would that work in the GUI?

The initial motivation of `ExecutionGraphInfo` was to encapsulate the issue having multiple `AccessExecutionGraph` instances in place for a given job. The `DefaultScheduler` only has one `AccessExecutionGraph` instance during the job's lifetime. The `AdaptiveScheduler` might come up with multiple instances of `AccessExecutionGraph` as it might run into global failovers. Instead of having multiple `ExecutionGraphInfo` instances, you could extend `ExecutionGraphInfo` to support multiple `AccessExecutionGraph` instances. Alternatively, one could come up with a `CompositeExecutionGraph` implementing `AccessExecutionGraph` which aggregates the underlying collection of `ArchivedExecutionGraphs`. That might reduce the amount of code changes that are necessary to be done in the REST handlers.

bytesandwich commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/SerializableAccessExecutionGraph.java:
```diff
@@ -0,0 +1,27 @@
+ * {@code SerializableAccessExecutionGraph} is a serializable version of {@link
+ * AccessExecutionGraph}.
+ */
+public interface SerializableAccessExecutionGraph extends AccessExecutionGraph, Serializable {}
```
This would now make 4 interfaces that are exclusively for 2 classes?

Interfaces: { `JobStatusProvider`, `AccessExecutionGraph`, `SerializableAccessExecutionGraph`, `ExecutionGraph` }
Classes: { `DefaultExecutionGraph`, `ArchivedExecutionGraph` }

Perhaps it is time to restrategize about refactoring this. `AccessExecutionGraph` is both a sort of "readonly" ExecutionGraph interface and also a placeholder for the `ArchivedExecutionGraph` but it can't really be both since an `ArchivedExecutionGraph` is strictly not an `ExecutionGraph`? I am not sure what to suggest and I can certainly see the logic in `AccessExecutionGraph` as a "readonly" `ExecutionGraph` interface.

XComp replied:
Thanks for sharing your perspective, @bytesandwich . You're right - the `ExecutionGraph`- and `Execution`-related classes/interfaces are up for discussion already for a while as they make testing individual code harder right now.

My proposal of introducing `SerializableAccessExecutionGraph` was solely based on considerations around [FLINK-21439](https://issues.apache.org/jira/browse/FLINK-21439) as stated in [my comment above](https://github.com/apache/flink/pull/15049#discussion_r598459965). But I am happy to revert the switch from `ArchivedExecutionGraph` to `AccessExecutionGraph` as the change might not be necessary depending on how we tackle `FLINK-21439` in the end.

XComp commented:
The replacement of `ArchivedExecutionGraph` by `AccessExecutionGraph`/`SerializableAccessExecutionGraph` was removed and the branch was rebased to the most recent version of `master`.

bytesandwich commented:
I think any more things I'd like to figure out can happen as I do FLINK-21439. Thank you for helping me understand @XComp @zentol .
